design_drafts/hardware_access.md

```diff
@@ -0,0 +1,160 @@
+We firstly describe how hardware resources are classified and loaded.
+Secondly, we provide a real-time safe way of accessing these resources through controllers.
+
+## Hardware Resources
+A hardware resource describes a physical component which is being considered through ros2-control.
+We hereby distinguish between three classes of hardware resources, namely Actuators, Sensor and System.
+Every individual hardware is loaded at runtime and thus allows a flexible and dynamic composition of the to be controlled setup.
+The hardware is composed uniquely through the URDF.
+
+**Joint**
+A joint is considered a logical component and is being actuated by at least one actuator (it might be generically under- or over-actuated depending on the actual hardware setup).
+The joint is meant to be abstraction layer between a high-level controller instance and the underlaying hardware.
+The abstraction is needed to shim over a potentially complex hardware setup in order to control a joint.
+A single joint might be controlled by multiple motors with a non-trivial transmission interface, yet a controller only really thinks in joint state values, not in transmission ratios.
+
+A joint is configured in conjunction to a hardware resource such as **Actuator**, **Sensor** or **System**.
+The joint component is configured through command and state interfaces.
+The command interfaces describe the value in which this joint can be controlled (e.g. effort or velocity) where as the state interfaces are considered the state feedback interfaces.
+An example URDF:
+```xml
+...
+<joint name="joint1">
+  <command_interface>velocity</command_interface>
+  <state_interface>position</state_interface>
+  <state_interface>velocity</state_interface>
+</joint>
+...
+```
+
+**Actuator**
+An actuator describes a single motor instance with at max 1DoF and is strictly tight to a **Joint**.
+It might hereby take a single command value for its appropriate mode of operation, such as a desired joint velocity or effort - in rare cases an actuator might actually take a precise joint position value.
+The implementation of this actuator might then convert the desired value into PWM or other hardware specific commands and control the hardware.
+Similarly, an actuator might provide state feedback.
+Depending the setup, the motor encoders might provide position, velocity, effort or current feedback.
+The implementation likewise might provide these values by connecting through the hardware driver.
+
+The URDF snippet for an actuator might look like the following:
+```xml
+<ros2_control name="my_simple_servo_motor" type="actuator">
+<hardware>
+  <class>simple_servo_motor_pkg/SimpleServoMotor</class>
+  <param name="serial_port">/dev/tty0</param>
+...
+</hardware>
+  <joint name="base_link">
+  <command_interface>position</command_interface>
+  <state_interface>position</state_interface>
+  <param name="max_value">1.57</param>
```

Comment:
We clarified the min-max notation in [the urdf examples](https://github.com/ros-controls/roadmap/blob/master/design_drafts/components_architecture_and_urdf_examples.md) where max and min are child nodes of the command interface, the state interface shouldn't have limits on it. The notation you use in this example looks simpler and I'd personally prefer it if it wasn't prone to misinterpretation. Better be explicit in this case.

In short,
```
      <commandInterfaceType name="position">
        <param name="min">-1</param>
        <param name="max">1</param>
      </commandInterfaceType>
```